README.md

## CytoCAD - Copy-number variation caller using low-depth whole-genome sequencing data
[![Build Status](https://app.travis-ci.com/cytham/cytocad.svg?branch=master)](https://app.travis-ci.com/github/cytham/cytocad)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/cytocad)](https://pypi.org/project/cytocad/)
[![PyPI versions](https://img.shields.io/pypi/v/cytocad)](https://pypi.org/project/cytocad/)
[![Conda](https://img.shields.io/conda/v/bioconda/cytocad)](https://anaconda.org/bioconda/cytocad)
[![Github release](https://img.shields.io/github/v/release/cytham/cytocad?include_prereleases)](../../releases)
[![PyPI license](https://img.shields.io/pypi/l/cytocad)](./LICENSE.txt)

<p align="center">
  <img src="https://user-images.githubusercontent.com/25361260/131828572-bfb57cf8-e9e2-4f8d-b200-5b4b5a5b8181.png" width="500" alt="accessibility text" align='center'>
</p>

CytoCAD is a bioinformatics tool for the identification of large genomic copy-number variation through coverage anomaly detection
 (CAD) using mapped whole-genome sequencing (WGS) data. It has been tested in low-depth (~8X) Oxford Nanopore WGS long-read
  data. Its output displays chromosome illustrations demarcating regions of copy-number gains (Red) or losses (Blue). The above illustration shows a loss of one chromosome 7 copy, a gain of one chromosome 21 copy, a partial duplication of both chromosome 8 copies, and a loss of one chromosome 17 short arm. It also has two X chromosomes and no Y chromosome, suggesting a female sex.

### Basic information:
* Takes as input a mapped whole-genome sequencing BAM file and output a BED file and a chromosome ideogram-like figure
* Uses [Ruptures](https://github.com/deepcharles/ruptures) python package for change point detection of read coverage data per
 chromosome 
* Uses [tagore](https://github.com/jordanlab/tagore) for chromosome ideogram illustrations

## Getting Started

### Quick run

```
cytocad [Options] sample.bam working_dir 
```

| Argument | Comment |
| :--- | :--- |
| sample.bam | Input mapped WGS BAM file |
| working_dir | Working directory |

#### Output
| Output file | Comment |
| :--- | :--- |
| ${sample}.ideo.svg | Chromosome ideogram produced by [tagore](https://github.com/jordanlab/tagore) |
| ${sample}.CNV.bed | BED file of chromosome regions with CNV |

For more information, see [wiki](https://github.com/cytham/cytocad/wiki).

### Operating system: 
* Linux (x86_64 architecture, tested in Ubuntu 16.04)  

### Installation:
There are three ways to install CytoCAD:
#### Option 1: Conda (Recommended)
```
# Installing from bioconda automatically installs all dependencies 
conda install -c bioconda cytocad
```
#### Option 2: PyPI (See dependencies below)
```
# Installing from PyPI requires own installation of dependencies, see below
pip install cytocad
```
#### Option 3: GitHub (See dependencies below)
```
# Installing from GitHub requires own installation of dependencies, see below
git clone https://github.com/cytham/cytocad.git 
cd cytocad
pip install .
```

### Installation of dependencies
* bedtools >=2.26.0
* samtools >=1.3.0
* rsvg-convert >=2.40.13

Please make sure each executable binary is in PATH.
##### 1. _bedtools_
Please visit [here](https://bedtools.readthedocs.io/en/latest/content/installation.html) for instructions to install.

##### 2. _samtools_
Please visit [here](http://www.htslib.org/download/) for instructions to install.

##### 3. _rsvg-convert_
```
sudo apt-get update
sudo apt-get install librsvg2-bin
```

## Versioning
See [CHANGELOG](./CHANGELOG.txt)

## Citation

Not available yet

## Author

* **Tham Cheng Yong** - [cytham](https://github.com/cytham)

## License

This project is licensed under GNU General Public License - see [LICENSE.txt](./LICENSE.txt) for details.

## Limitations
* Chromosome pairs illustrated by tagore may resemble sister chromatids, but they are in fact homologous pairs
* Phasing of CNVs for each chromosome homologous pair is not yet possible.
* The default minimum size of detectable CNV is about 500 kb. It can be adjusted by the 'interval' and 'rolling' parameters
 following the equation: minimum size ~= interval*rolling 
* Other chromosomal structural variations, such as inversions, have to be detected by other tools, such as [NanoVar](https://github.com/cytham/nanovar). NanoVar has incorporated CytoCAD in its pipeline from version 1.4.0 onwards.


CHANGELOG.txt
CytoCAD Changelog


Release Summary:


Version 1.0.3 - Oct 8, 2021
    * Changed coverage plot y axis to copy number


Version 1.0.2 - Sept 8, 2021
    * Add ruptures in Bioconda recipe
    
    
Version 1.0.1 - Sept 1, 2021
    * Scaled genomic coverage to fix high coverage bug
    * Extended chr Y end filter region


Version 1.0.0 - May 29, 2021
    * Initial release


cytocad/input.py
"""
Functions for parsing and verifying input parameters.

Copyright (C) 2021 Tham Cheng Yong

This file is part of CytoCAD.

CytoCAD is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CytoCAD is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CytoCAD.  If not, see <https://www.gnu.org/licenses/>.
"""


import sys
import argparse
from cytocad import __version__


# Parse input
def input_parser(args=sys.argv[1:]):
    parser = argparse.ArgumentParser(description="CytoCAD is a tool for discovering large genomic copy-number variation \
through coverage anormaly detection (CAD) using low-depth whole-genome sequencing data.",
                                     formatter_class=argparse.RawTextHelpFormatter, usage=msg())

    parser.add_argument("input", type=str,
                        metavar="[BAM]",
                        help="""path to mapped BAM file.
Format: .bam""")

    parser.add_argument("dir", type=str,
                        metavar="[work_directory]",
                        help="""path to work directory. Directory will be created 
if it does not exist.""")

    parser.add_argument("-b", "--build", type=str, metavar="str",
                        default='hg38',
                        help="""build version of human reference genome assembly [hg38]""")

    parser.add_argument("-c", "--colors", nargs='+', metavar="hex_color",
                        default=None,
                        help="""hex color for neutral, gain, and loss CNVs on chromosome
ideograms respectively separated by space ['#a6a6a6' '#990000' '#000099']""")

    parser.add_argument('-f', '--format', type=str, metavar='[png/pdf]',
                        default='png',
                        help="Output format of chromosome illustration figure [png]")

    parser.add_argument("-i", "--interval", type=int, metavar="int",
                        default=50000,
                        help="""spread between each point in a chromosome where "
coverage is enquired, in bp. Minimum CNV sensitive 
detection size ~= interval*rolling [50000]""")

    parser.add_argument("-j", "--buffer", type=int, metavar="int",
                        default=10,
                        help="buffer window size of each point, in bp [10]")

    parser.add_argument("-r", "--rolling", type=int, metavar="int",
                        default=10,
                        help="rolling mean window size [10]")

    parser.add_argument("-p", "--penalty", type=int, metavar="int",
                        default=500,
                        help="""Linear kernel penalty value for change 
point detection using Ruptures [500]""")

    parser.add_argument("-s", "--scale", type=float, metavar="float",
                        default=0.25,
                        help="""proportion of mean coverage to be used for 
buffering to call hetero- and homozygous CNVs 
(E.g. a heterozygous loss is where a coverage (c) 
satisfies: mean-mean*scale <= c < mean+mean*scale [0.25]""")

    parser.add_argument("--add_plots", action='store_true',
                        help="output additional coverage plots in 'fig' directory")

    parser.add_argument("--debug", action='store_true',
                        help="run in debug mode")

    parser.add_argument("-v", "--version", action='version',
                        version=__version__,
                        help="show version and exit")

    parser.add_argument("-q", "--quiet", action='store_true',
                        help="hide verbose")

    args = parser.parse_args(args)
    return args


# Custom usage message
def msg():
    return "cytocad [options] [BAM] [WORK_DIRECTORY]"


cytocad/ideogram.py
"""
A wrapper for tagore.

Copyright (C) 2021 Tham Cheng Yong

This file is part of CytoCAD.

CytoCAD is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CytoCAD is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CytoCAD.  If not, see <https://www.gnu.org/licenses/>.
"""


import os
# import pickle
# import cytocad
import logging
from subprocess import Popen, PIPE, STDOUT
# from cytocad.tagore import draw


# Wrapper to visualize CNVs using tagore
def tagore_wrapper(tagout, sample_name, wk_dir, ref_build, oformat):
    # Write tagore BED to file
    out_path = os.path.join(wk_dir, sample_name + '.tagore.bed')
    outwrite = open(out_path, 'w')
    _ = outwrite.write('\n'.join(tagout))
    outwrite.close()
    prefix = os.path.join(wk_dir, sample_name + '.ideo')
    process = Popen(['tagore', '-i', out_path, '-p', prefix, '-b', ref_build, '-ofmt', oformat, '-f'],
                    universal_newlines=True, stdout=PIPE, stderr=STDOUT)
    with process.stdout:
        log_subprocess(process.stdout)
    exitcode = process.wait()
    if exitcode != 0:
        logging.critical("Error: tagore failed")
        raise Exception("Error: tagore failed, see log")


def log_subprocess(out):
    for line in iter(out.readline, ''):
        if line != '\n':
            logging.debug(line.strip())


# Wrapper to visualize CNVs using tagore standalone
# def tagore_standalone(tagout, sample_name, wk_dir, ref_build):
#     # Write tagore BED to file
#     out_path = os.path.join(wk_dir, sample_name + '.tagore.bed')
#     outwrite = open(out_path, 'w')
#     _ = outwrite.write('\n'.join(tagout))
#     outwrite.close()
#     prefix = os.path.join(wk_dir, sample_name + '.ideo')
#     tag_args = TagoreArgs(out_path, prefix, ref_build)
#     data_dir = os.path.join(os.path.dirname(cytocad.__file__), 'data')
#     base_path = os.path.join(data_dir, 'base.svg.p')
#     base = open(base_path, 'rb')
#     __head__, __tail__ = pickle.load(base)
#     draw(tag_args, __head__, __tail__)


# Format tagore arguments
# class TagoreArgs:
#
#     def __init__(self, bed_path, sample_prefix, ref_build):
#         self.input = bed_path
#         self.prefix = sample_prefix
#         self.build = ref_build


cytocad/depth_limit.py
"""
Functions to calculate the upper limit for depth of coverage.

Copyright (C) 2021 Tham Cheng Yong

This file is part of CytoCAD.

CytoCAD is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CytoCAD is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CytoCAD.  If not, see <https://www.gnu.org/licenses/>.
"""


import os
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker
from pybedtools import BedTool
from scipy.interpolate import make_interp_spline


# Generate upper overlap limit, depth of coverage, and coverage curve plot
def ovl_upper(total_gsize, contig_len_dict, subdata, wk_dir, cov_plots):
    bed_str = normalbed(subdata)
    bed = BedTool(bed_str, from_string=True)
    bed = bed.sort()
    n = ngenerate(total_gsize)
    gsize_path = make_gsize(contig_len_dict, wk_dir)
    x = BedTool()
    ran_bed = x.random(l=100, n=n, seed=3, g=gsize_path)
    os.remove(gsize_path)
    ran_bed = ran_bed.sort()
    insect = ran_bed.intersect(bed, wa=True, wb=True)
    cov_list = []
    for line in insect:
        line = list(line)
        cov_list.append(line[3])
    cov_dict = {}
    for key in cov_list:
        try:
            cov_dict[key] += 1
        except KeyError:
            cov_dict[key] = 1
    data = []
    for key in cov_dict:
        data.append(cov_dict[key])
    data = [float(i) for i in data]
    zerolist = [0.0]*(n-len(data))
    data2 = data + zerolist
    med = np.median(data2)
    medad = mad(data2)
    maxovl = max(round((medad * 3) + med, 1), 10)  # minimum overlap threshold is set at 10
    if cov_plots:
        curve(data2, n, round((medad*6) + med, 0), wk_dir)
    return maxovl


# Make genome size file
def make_gsize(contig_len_dict, wk_dir):
    path = os.path.join(wk_dir, 'genome.sizes')
    data = open(path, 'w')
    tmp = []
    for contig in contig_len_dict:
        # checkcontigname(contig)
        tmp.append(contig + '\t' + str(contig_len_dict[contig]))
    data.write('\n'.join(tmp) + '\n')
    data.close()
    return path


# Generate number of genomic points
def ngenerate(gsize):
    if gsize > 10000:
        return 10000
    else:
        return int(gsize*0.9)


# Function to generate bed file from subdata
def normalbed(subdata):
    totalbed = []
    for line in subdata:
        coord1 = int(line.split('\t')[1])
        coord2 = int(line.split('\t')[1]) + int(line.split('\t')[2])
        totalbed.append(line.split('\t')[0] + '\t' + str(coord1) + '\t' + str(coord2) + '\t' + line.split('\t')[4])
    bed_str = '\n'.join(totalbed)
    return bed_str


# Median absolute deviation
def mad(x):
    b = 1.4826
    return b*np.mean(abs(x-np.median(x)))


# Plot curve
def curve(data, n, upper_limit, wk_dir):
    c = range(max(int(upper_limit), 10))
    p = []
    for i in c:
        p.append(data.count(i))
    p.append(n - sum(p))
    y = np.array([(float(z)/n) for z in p])
    # theoretical = [0.0915, 0.0441, 0.1032, 0.1498, 0.1739, 0.1626,
    # 0.1132, 0.0808, 0.0412, 0.0247, 0.0097, 0.0028, 0.0015,
    # 0.0006, 0.0002, 0.0002, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
    c = np.array(range(max(int(upper_limit), 10)+1))
    xnew = np.linspace(c.min(), c.max(), 100)
    spl = make_interp_spline(c, y)
    smooth = spl(xnew)
    params = {'axes.labelsize': 14, 'axes.titlesize': 17, 'legend.fontsize': 10, 'xtick.labelsize': 12, 'ytick.labelsize': 12,
              'font.family': 'Arial, Helvetica, sans-serif'}
    matplotlib.rcParams.update(params)
    fig = plt.figure(figsize=(8, 6))
    fig.patch.set_facecolor('#f6f7f9')
    ax = fig.add_subplot(111)
    ax.plot(xnew, smooth, color='#403f7d', linewidth=2.0)
    ax.set_facecolor('#ebebff')
    plt.text(int(upper_limit), y[-1], '>=' + str(int(upper_limit)))
    ax.grid(color='w', linestyle='-', linewidth=1)
    vals = ax.get_yticks().tolist()
    ax.yaxis.set_major_locator(mticker.FixedLocator(vals))
    ax.set_yticklabels(['{:,.1%}'.format(x) for x in vals])
    plt.ylabel('Percentage')
    plt.xlabel('Depth of coverage')
    ax.legend(['Input sequencing data'], loc='upper right', ncol=1, fancybox=True)
    plt.savefig(os.path.join(wk_dir, 'fig', 'total_depth_of_coverage.svg'),
                bbox_inches='tight',
                dpi=100,
                facecolor=fig.get_facecolor(),
                edgecolor='none')


cytocad/change_detection.py
"""
Detect coverage changes by change point detection

Copyright (C) 2021 Tham Cheng Yong

This file is part of CytoCAD.

CytoCAD is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CytoCAD is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CytoCAD.  If not, see <https://www.gnu.org/licenses/>.
"""


import os
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from pybedtools import BedTool
from collections import OrderedDict, defaultdict
import ruptures as rpt
import cytocad
from cytocad.depth_limit import normalbed


# Coverage anomaly detection
def cad(
        subdata,
        mean_cov,
        upper_cov,
        sample_name,
        ref_build='hg38',
        interval=50000,
        interval_buf=10,
        rolling_size=10,
        penalty=500,  # 500
        zygo_scale=0.25,
        cov_plots=False,
        wk_dir='./',
        colors=None  # [Neutral, Gain, Loss]
):
    # Define colors
    if colors is None:
        neutral_color, gain_color, loss_color = ['#a6a6a6', '#990000', '#000099']
    else:
        if len(colors) == 3:
            neutral_color, gain_color, loss_color = colors
        else:
            raise Exception('Error: The variable "colors" has to be a list of length 3, indicating the hex color of Neutral, '
                            'Gain and Loss CNVs in this sequential order.')

    # Define data directory
    data_dir = os.path.join(os.path.dirname(cytocad.__file__), 'data')

    # Define file paths according to reference build
    if ref_build == 'hg38':
        filter_path = os.path.join(data_dir, 'hg38_curated_filter_main_arte.bed')
        ideo_path = os.path.join(data_dir, 'hg38_ucsc_ideogram.bed')
        main_chr_path = os.path.join(data_dir, 'hg38_sizes_main.bed')
    else:
        raise Exception("Error: Reference genome build %s is not recognised. CytoCAD only supports build hg38." % ref_build)

    # Define other variables
    buf = mean_cov * zygo_scale
    het_loss = mean_cov / 2
    hom_loss = 0
    het_gain = mean_cov + mean_cov / 2
    hom_gain = mean_cov * 2

    # Make filtered-region BedTool object
    filter_bed = BedTool(filter_path)

    # Create intervals of genomic coordinates for each chromosome
    main_chr = set()
    chr_range = OrderedDict()
    with open(main_chr_path) as f:
        for line in f:
            chrm, start, end = line.split('\t')
            chr_range[chrm] = range(int(start), int(end), interval)
            main_chr.add(chrm)
    sort_chr = list(main_chr)
    sort_chr.sort()

    # Make read alignment BedTool object
    bed_str = normalbed(subdata)
    bed = BedTool(bed_str, from_string=True)
    bed = bed.sort()

    # For each chromosome, analyse read coverage by BED intersect
    data = defaultdict(list)
    region = defaultdict(list)
    chrx_avg = 0
    chry_avg = 0
    for chromo in chr_range:
        # print('Processing ' + chromo)
        cov_dict = {}
        region_dict = {}
        string = ''
        for i in chr_range[chromo]:
            string += chromo + '\t' + str(max(i - interval_buf, 1)) + '\t' + str(i + interval_buf) + '\n'
            cov_dict[chromo + '-' + str(max(i - interval_buf, 1))] = 0
            region_dict[chromo + '-' + str(max(i - interval_buf, 1))] = max(i, 1)
        interval_bed = BedTool(string, from_string=True)
        interval_bed = interval_bed.sort()
        intersect1 = interval_bed.intersect(filter_bed, wa=True, v=True)  # Remove gap regions, prevent false amplifications
        intersect2 = intersect1.intersect(bed, wa=True, wb=True)  # Intersect with read alignment data
        intersect3 = intersect1.intersect(bed, wa=True, v=True)  # Not intersected with read alignment data
        # Get number of read coverage at each interval
        cov_list = []
        for line in intersect2:
            line = list(line)
            cov_list.append(line[0] + '-' + line[1])
        for key in cov_list:
            cov_dict[key] += 1
        if chromo == 'chrX':
            total = 0
            nocov = 0
            for key in set(cov_list):
                total += cov_dict[key]
            for _ in intersect3:
                nocov += 1
            chrx_avg = total / (len(set(cov_list)) + nocov)
            # print('Chrx avg: ' + str(chrx_avg))
        elif chromo == 'chrY':
            total = 0
            nocov = 0
            for key in set(cov_list):
                total += cov_dict[key]
            for _ in intersect3:
                nocov += 1
            chry_avg = total / (len(set(cov_list)) + nocov)
            # print('Chry avg: ' + str(chry_avg))
        # For each region falling in filter bed, assign mean coverage
        intersect4 = interval_bed.intersect(filter_bed, wa=True)
        for line in intersect4:
            line = list(line)
            if chromo == 'chrX':
                cov_dict[line[0] + '-' + line[1]] = chrx_avg
            elif chromo == 'chrY':
                cov_dict[line[0] + '-' + line[1]] = chry_avg
            else:
                cov_dict[line[0] + '-' + line[1]] = mean_cov
        for key in cov_dict:
            data[key.split('-')[0]].append(min(cov_dict[key], upper_cov))  # coverage counts
            region[key.split('-')[0]].append(region_dict[key]/1000000)  # genomic coordinate

    # Process ideogram coordinates
    ideo_dict = {}
    with open(ideo_path) as f:
        for i in f:
            if not i.startswith('#'):
                if i.split('\t')[0] not in ideo_dict:
                    ideo_dict[i.split('\t')[0]] = {}
                ideo_dict[i.split('\t')[0]][str(i.split('\t')[1]) + '-' + str(i.split('\t')[2])] = i.split('\t')[3]

    # Process by eight chromosomes cycle
    groups = [[0, 8], [8, 16], [16, 24]]
    tagout = ['#chr\tstart\tstop\tfeature\tsize\tcolor\tchrCopy']
    out = []
    cycle = 1
    for g in groups:
        if cov_plots:
            fig = plt.figure()
        n = 1
        for chromo in sort_chr[g[0]:g[1]]:
            d = {'x': region[chromo], 'y': data[chromo]}
            df = pd.DataFrame(d)
            xcoord = df.x.to_numpy()
            rolling_mean = df.y.rolling(window=rolling_size).mean()
            signal = np.array(rolling_mean)
            signal[np.isnan(signal)] = mean_cov
            # Scale coverage for consistent change detection
            if chromo == 'chrX':
                signal_scaled = signal/chrx_avg * 8
                signal_plot = signal/chrx_avg
            elif chromo == 'chrY':
                signal_scaled = signal/chry_avg * 8
                signal_plot = signal/chry_avg
            else:
                signal_scaled = signal/mean_cov * 8
                signal_plot = signal/mean_cov * 2
            algo = rpt.KernelCPD(kernel="linear", min_size=2).fit(signal_scaled)
            result = algo.predict(pen=penalty)
            _span = []
            # if len(result) % 2 != 0:
            _result = [0] + result
            for i in range(len(_result) - 1):
                _span.append([_result[i], _result[i + 1]])
            # else:
            #     _result = [0] + result
            #     for i in range(len(_result) - 1):
            #         _span.append([_result[i], _result[i + 1]])
            ideo = {}
            cov = {}
            coord = {}
            for i in _span:
                span_ideo = []
                left = region[chromo][i[0]] * 1000000
                right = region[chromo][i[1] - 1] * 1000000
                for j in ideo_dict[chromo]:
                    if int(j.split('-')[0]) < left <= int(j.split('-')[1]):
                        span_ideo.append(ideo_dict[chromo][j])
                    if int(j.split('-')[0]) < right <= int(j.split('-')[1]):
                        span_ideo.append(ideo_dict[chromo][j])
                        break
                ideo[str(i[0]) + '-' + str(i[1])] = span_ideo
                size = len(range(i[0], i[1] - 1))
                s = 0
                for d in range(i[0], i[1] - 1):
                    s += data[chromo][d]
                cov[str(i[0]) + '-' + str(i[1])] = round(s / size, 3)
                coord[str(i[0]) + '-' + str(i[1])] = [int(left), int(right)]
            label = {}
            if chromo == 'chrX':
                if mean_cov - buf <= chrx_avg:  # XX
                    signal_plot *= 2
                    for i in cov:
                        if het_loss - buf <= cov[i] < het_loss + buf:
                            label[i] = 'loss-hetero'
                        elif hom_loss <= cov[i] < hom_loss + buf:
                            label[i] = 'loss-homo'
                        elif het_gain - buf <= cov[i] < het_gain + buf:
                            label[i] = 'gain-hetero'
                        elif hom_gain - buf <= cov[i]:
                            label[i] = 'gain-homo'
                        else:
                            label[i] = 'neutral-double'
                else:  # X or no X
                    for i in cov:
                        if hom_loss <= cov[i] < hom_loss + buf:
                            label[i] = 'loss-single'
                        elif mean_cov / 2 + buf <= cov[i]:
                            label[i] = 'gain-single'
                        else:
                            label[i] = 'neutral-single'
            elif chromo == 'chrY':
                if mean_cov / 2 - buf <= chry_avg:  # Y
                    for i in cov:
                        if hom_loss <= cov[i] < hom_loss + buf:
                            label[i] = 'loss-single'
                        elif mean_cov / 2 + buf <= cov[i]:
                            label[i] = 'gain-single'
                        else:
                            label[i] = 'neutral-single'
                else:  # no Y
                    for i in cov:
                        label[i] = 'nil-nil'
            else:
                for i in cov:
                    if het_loss - buf <= cov[i] < het_loss + buf:
                        label[i] = 'loss-hetero'
                    elif hom_loss <= cov[i] < hom_loss + buf:
                        label[i] = 'loss-homo'
                    elif het_gain - buf <= cov[i] < het_gain + buf:
                        label[i] = 'gain-hetero'
                    elif hom_gain - buf <= cov[i]:
                        label[i] = 'gain-homo'
                    else:
                        label[i] = 'neutral-double'
            for i in cov:
                copy = label[i].split('-')[0]
                zygo = label[i].split('-')[1]
                if copy == 'neutral':
                    if zygo == 'double':
                        tagout.append(
                            chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t0\t1\t' + neutral_color + '\t1'
                        )
                        tagout.append(
                            chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t0\t1\t' + neutral_color + '\t2'
                        )
                    elif zygo == 'single':
                        tagout.append(
                            chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t0\t1\t' + neutral_color + '\t1'
                        )
                elif copy == 'nil':
                    continue
                else:
                    if copy == 'gain':
                        color = gain_color
                    else:  # loss
                        color = loss_color
                    if zygo == 'hetero':
                        tagout.append(
                            chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t0\t1\t' + color + '\t1'
                        )
                        tagout.append(
                            chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t0\t1\t' + neutral_color + '\t2'
                        )
                    elif zygo == 'homo':
                        tagout.append(
                            chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t0\t1\t' + color + '\t1'
                        )
                        tagout.append(
                            chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t0\t1\t' + color + '\t2'
                        )
                    elif zygo == 'single':
                        tagout.append(
                            chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t0\t1\t' + color + '\t1'
                        )
                    out.append(
                        chromo + '\t' + str(coord[i][0]) + '\t' + str(coord[i][1]) + '\t' + ideo[i][0] + ';' + ideo[i][1] + '\t' +
                        str(round(cov[i], 1)) + '\t' + copy + '\t' + zygo
                    )
            if cov_plots:
                span = []
                if len(result) % 2 != 0:
                    result.append(0)
                    result_iter = iter(result)
                    for i in result_iter:
                        span.append([i, next(result_iter)])
                    _ = span.pop(-1)
                else:
                    result_iter = iter(result)
                    for i in result_iter:
                        span.append([i, next(result_iter)])
                ax = fig.add_subplot(2, 4, n)
                ax.plot(xcoord, signal_plot, label='SMA', color='red', alpha=0.8)
                for p in span:
                    ax.axvspan(region[chromo][p[0] - 1], region[chromo][p[1] - 1], alpha=0.3, color='blue')
                ax.set_title(chromo)
                ax.set_ylim(0, 5)
                ax.set_yticks(np.arange(0, 5, 1))
                ax.yaxis.grid(True)
                n += 1
        if cov_plots:
            n = 1
            fig.add_subplot(111, frame_on=False)
            plt.tick_params(labelcolor="none", bottom=False, left=False)
            plt.xlabel('Coordinate (MB)')
            plt.ylabel('Copy number')
            plt.tight_layout()
            fig_out_path = os.path.join(wk_dir, 'fig', sample_name + '_cov' + str(cycle) + '.svg')
            plt.savefig(fig_out_path,
                        dpi=100
                        )
            cycle += 1
    return out, tagout
